[Community Pipelines]Accelerate inference of stable diffusion by IPEX on CPU (#3105)
huggingface/diffusers: examples/community/README.md

| UnCLIP Image Interpolation Pipeline | Diffusion Pipeline that allows passing two images/image_embeddings and produces images while interpolating between their image-embeddings | [UnCLIP Image Interpolation Pipeline](#unclip-image-interpolation-pipeline)                   | -                                                                                                                                                                                                                  | [Naga Sai Abhinay Devarinti](https://github.com/Abhinay1997/) | 
| DDIM Noise Comparative Analysis Pipeline | Investigating how the diffusion models learn visual concepts from each noise level (which is a contribution of [P2 weighting (CVPR 2022)](https://arxiv.org/abs/2204.00227)) | [DDIM Noise Comparative Analysis Pipeline](#ddim-noise-comparative-analysis-pipeline) | - |[Aengus (Duc-Anh)](https://github.com/aengusng8) |
| CLIP Guided Img2Img Stable Diffusion Pipeline | Doing CLIP guidance for image to image generation with Stable Diffusion | [CLIP Guided Img2Img Stable Diffusion](#clip-guided-img2img-stable-diffusion) | - | [Nipun Jindal](https://github.com/nipunjindal/) | 

> | Stable Diffusion IPEX Pipeline | Accelerate Stable Diffusion inference pipeline with BF16/FP32 precision on Intel CPUs by [IPEX](https://github.com/intel/intel-extension-for-pytorch) | [Stable Diffusion on IPEX](#stable-diffusion-on-ipex) | - | [Yingjie Han](https://github.com/yingjie-han/) | 

Reviewer: ```suggestion
| Stable Diffusion IPEX Pipeline | Accelerate Stable Diffusion inference pipeline with BF16/FP32 precision on Intel Xeon CPUs with [IPEX](https://github.com/intel/intel-extension-for-pytorch) | [Stable Diffusion on IPEX](#stable-diffusion-on-ipex) | - | [Yingjie Han](https://github.com/yingjie-han/) | 
```

Output Image

![img2img_clip_guidance](https://huggingface.co/datasets/njindal/images/resolve/main/clip_guided_img2img.jpg)

### Stable Diffusion on IPEX

> This diffusion pipeline aims to accelarate the inference of Stable-Diffusion on Intel Xeon CPUs with BF16/FP32 precision by [IPEX](https://github.com/intel/intel-extension-for-pytorch).

Reviewer: ```suggestion
This diffusion pipeline aims to accelarate the inference of Stable-Diffusion on Intel Xeon CPUs with BF16/FP32 precision using [IPEX](https://github.com/intel/intel-extension-for-pytorch).
```

> To use this pipeline, You need to:

Reviewer: ```suggestion
To use this pipeline, you need to:
```

1. Install [IPEX](https://github.com/intel/intel-extension-for-pytorch)

> **Note:** For each PyTorch release, there is a corresponding release of the IPEX. Here are the mapping relationship.It is recommanded to install Pytorch/IPEX2.0 to get the best performance.

Reviewer: ```suggestion
**Note:** For each PyTorch release, there is a corresponding release of the IPEX. Here is the mapping relationship. It is recommended to install Pytorch/IPEX2.0 to get the best performance.
```

|PyTorch Version|IPEX Version|
|--|--|
|[v2.0.\*](https://github.com/pytorch/pytorch/tree/v2.0.1 "v2.0.1")|[v2.0.\*](https://github.com/intel/intel-extension-for-pytorch/tree/v2.0.100+cpu)|
|[v1.13.\*](https://github.com/pytorch/pytorch/tree/v1.13.0 "v1.13.0")|[v1.13.\*](https://github.com/intel/intel-extension-for-pytorch/tree/v1.13.100+cpu)|

> You can use normal pip command to install IPEX with the latest version.

Reviewer: ```suggestion
You can simply use pip to install IPEX with the latest version.
```

Output Image

![img2img_clip_guidance](https://huggingface.co/datasets/njindal/images/resolve/main/clip_guided_img2img.jpg)

### Stable Diffusion on IPEX

This diffusion pipeline can accelarate the inference of Stable-Diffusion on Intel CPUs with BF16/FP32 precision by [IPEX](https://github.com/intel/intel-extension-for-pytorch).

To use this pipeline, You need to:
1. Install [IPEX](https://github.com/intel/intel-extension-for-pytorch)
```python
python -m pip install intel_extension_for_pytorch

> ```

Reviewer: My understanding is that IPEX requires specific versions of PyTorch (1.13.1?), so I'd also mention it here, particularly because PyTorch 2 is now the default and more performant in other platforms.

Author: Pytorch/IPEX2.0 is working. It is recommended to use Pytorch/IPEX2.0 to get the best performance boost of flash attention optimization. It can also work on Pytorch/IPEX 1.13. I have added the information about the version.

> 2. After pipeline initialization, prepare_for_ipex() should be called to enable IPEX accelaration.

Reviewer: ```suggestion
2. After pipeline initialization, `prepare_for_ipex()` should be called to enable IPEX accelaration.
```

```python
pipe = DiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", custom_pipeline="stable_diffusion_ipex")

> pipe.prepare_for_ipex(prompt,infer_type='bf16')

Reviewer: ```suggestion
pipe.prepare_for_ipex(prompt, infer_type="bf16")
```

```

> Other usage of this ipex pipeline is same as the default stable diffusion pipeline.

Reviewer: ```suggestion
Then you can use the ipex pipeline in exactly the same way as the default stable diffusion pipeline.
```

I'd include a short inference snippet to that effect.

> Following code compares the performance of original stable diffusion pipeline with ipex pipeline.

Reviewer: ```suggestion
The following code compares the performance of the original stable diffusion pipeline with the ipex-optimized pipeline.
```

```python
import torch
import intel_extension_for_pytorch as ipex
from diffusers import StableDiffusionPipeline
import time

prompt = "sailing ship in storm by Rembrandt"
model_id = "runwayml/stable-diffusion-v1-5"

> #Help function for time evaluation

Reviewer: ```suggestion
# Helper function for time evaluation
```

def elapsed_time(pipeline, nb_pass=3, num_inference_steps=20):
    # warmup
    for _ in range(2):
        images = pipeline(prompt, num_inference_steps=num_inference_steps).images
    #time evaluation
    start = time.time()
    for _ in range(nb_pass):
        pipeline(prompt, num_inference_steps=num_inference_steps)
    end = time.time()
    return (end - start) / nb_pass

##############     bf16 inference performance    ###############

> #1.IPEX Pipeline initialization

Reviewer: ```suggestion
# 1. IPEX Pipeline initialization
```

pipe = DiffusionPipeline.from_pretrained(model_id, custom_pipeline="stable_diffusion_ipex")
pipe.prepare_for_ipex(prompt,infer_type='bf16')

> #2.Original Pipeline initialization

Reviewer: ```suggestion
# 2. Original Pipeline initialization
```

pipe2 = StableDiffusionPipeline.from_pretrained(model_id)

> #3.Compare performance between Original Pipeline and IPEX Pipeline

Reviewer: ```suggestion
# 3. Compare performance between Original Pipeline and IPEX Pipeline
```

> with torch.no_grad(), torch.cpu.amp.autocast(enabled=True, dtype=torch.bfloat16):

Reviewer: `torch.no_grad()` is not required for pipeline inference. Is `autocast` required for `bfloat16`? In the original diffusers code, it is _not_ required for `float16`.

Author: yes, autocast required for bfloat16.

    latency = elapsed_time(pipe)
    print("Latency of StableDiffusionIPEXPipeline--bf16", latency)
    latency = elapsed_time(pipe2)
    print("Latency of StableDiffusionPipeline--bf16",latency)

##############     fp32 inference performance    ###############

> #1.IPEX Pipeline initialization

Reviewer: ```suggestion
# 1. IPEX Pipeline initialization
```

pipe3 = DiffusionPipeline.from_pretrained(model_id, custom_pipeline="stable_diffusion_ipex")
pipe3.prepare_for_ipex(prompt,infer_type='fp32')

> #2.Original Pipeline initialization

Reviewer: ```suggestion
# 2. Original Pipeline initialization
```

pipe4 = StableDiffusionPipeline.from_pretrained(model_id)

> #3.Compare performance between Original Pipeline and IPEX Pipeline

Reviewer: ```suggestion
# 3. Compare performance between Original Pipeline and IPEX Pipeline
```

> with torch.no_grad():

Reviewer: `torch.no_grad()` is not required for pipeline inference.

Author: I removed it.
